### PR TITLE
Rsdk-3264: RTK tests 

### PR DESCRIPTION
**Pull request by susmitaSanyal**
This PR includes tests for rtk over i2c and serial and some cleanups.

### CODE REVIEW

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+
+	// Normal position
+	loc1, alt1, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc1, test.ShouldResemble, geo.NewPoint(40.7, -73.98))
+	test.That(t, alt1, test.ShouldEqual, alt)
+
+	speed1, err := g.LinearVelocity(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, speed1.Y, test.ShouldEqual, speed)
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:125]: could also test that x and z are still zero

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+	testRoverName   = "testRover"
+	testStationName = "testStation"
+	testBoardName   = "board1"
+	testBusName     = "bus1"
+	testi2cAddr     = 44
+)
+
+func TestValidateRTK(t *testing.T) {
+	path := "path"
+	fakecfg := &Config{
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:32]: could we also add a test for a config that passes validation?

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+	test.That(t, speed2.Y, test.ShouldEqual, speed)
+
+	// Position with NaN values.
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(math.NaN(), math.NaN()), math.NaN(), nil
+	}
+
+	loc3, alt3, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, math.IsNaN(loc3.Lat()), test.ShouldBeTrue)
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:148]: isn't the rtk supposed to return last known position if the position is 0 or NaN?

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+	// create new ntrip client and connect
+	err := g.connect("invalidurl", username, password, 10)
+	g.ntripClient = makeMockNtripClient()
+
+	test.That(t, err, test.ShouldNotBeNil)
+
+	err = g.connect(url, username, password, 10)
+	test.That(t, err, test.ShouldBeNil)
+
+	err = g.getStream("", 10)
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:93]: what is this testing exactly? Checking for the specific error returned in the scenario you're testing could also make this clearer

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 5]
```diff
@@ -0,0 +1,189 @@
...
+	testi2cAddr     = 44
+)
+
+func TestValidateRTK(t *testing.T) {
+	path := "path"
+	fakecfg := &Config{
+		Board:                testBoardName,
+		I2CBus:               testBusName,
+		I2CAddr:              testi2cAddr,
+		I2CBaudRate:          4400,
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:36]: I would change to a valid baud rate just so theres no confusion, also might be worth adding in the check for a valid baud rate in the validate using the utils helper function

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 6]
```diff
@@ -0,0 +1,189 @@
...
+		cancelFunc: cancelFunc,
+		logger:     logger,
+	}
+
+	mockSensor := &CustomMovementSensor{
+		MovementSensor: &fake.MovementSensor{},
+	}
+
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(40.7, -73.98), 50.5, nil
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:112]: I think it's safer to initialize this point as a variable outside and have this function return that variable? You can then check `loc1` against that variable in your test condition

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 7]
```diff
@@ -0,0 +1,189 @@
...
+		return c.PositionFunc(ctx, extra)
+	}
+	// Fallback to the default implementation if PositionFunc is not set.
+	return c.MovementSensor.Position(ctx, extra)
+}
+
+// mock ntripinfo client.
+func makeMockNtripClient() *rtk.NtripInfo {
+	return &rtk.NtripInfo{}
+}
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:189]: should there be a test for the constructor?

**Review by oliviamiller — COMMENTED** [review]
just had some questions/suggestions

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 8]
```diff
@@ -0,0 +1,189 @@
...
+	"github.com/edaniels/golog"
+	geo "github.com/kellydunn/golang-geo"
+	"go.viam.com/test"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/movementsensor/fake"
+	rtk "go.viam.com/rdk/components/movementsensor/rtkutils"
+)
+
+var (
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:17]: I think it's clearer to define these variables in `TestPosition` instead since that's where they are being used

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go (hunk 3 above)
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:148]: Great question, it is it handled in nmea. And (0,0) is a valid position. NaN is also valid given that rtk received it from nmea meaning something didn't workout and instead of returning (0,0) nmea returned NaN()

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+	test.That(t, err, test.ShouldNotBeNil)
+
+	err = g.connect(url, username, password, 10)
+	test.That(t, err, test.ShouldBeNil)
+
+	err = g.getStream("", 10)
+	test.That(t, err, test.ShouldNotBeNil)
+}
+
+func TestPostion(t *testing.T) {
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:97]: this is testing more than `Position`, it's also testing `LinearVelocity`. Maybe the linear velocity tests should be in their own test function?

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 10]
```diff
@@ -0,0 +1,189 @@
...
+		logger:     logger,
+	}
+	g.ntripClient = &rtk.NtripInfo{}
+	g.nmeamovementsensor = &fake.MovementSensor{}
+
+	err := g.Close(ctx)
+	test.That(t, err, test.ShouldBeNil)
+}
+
+type CustomMovementSensor struct {
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:173]: why not just use `inject.MovementSensor`? Are we trying to move away from that?

File: components/movementsensor/gpsrtkserial/gpsrtkserial_test.go [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+	"github.com/edaniels/golog"
+	geo "github.com/kellydunn/golang-geo"
+	"go.viam.com/test"
+	"go.viam.com/utils"
+
+	"go.viam.com/rdk/components/movementsensor/fake"
+	rtk "go.viam.com/rdk/components/movementsensor/rtkutils"
+)
+
```
Comment (gvaradarajan) [inline components/movementsensor/gpsrtkserial/gpsrtkserial_test.go:16]: all the same comments/questions I had in `gpsrtkpmtk_test.go` apply to this file as well

**Review by gvaradarajan — CHANGES_REQUESTED** [review]

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go (hunk 3 above)
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:152]: it's the parser that returns NaN()

Comment (oliviamiller) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:152]: ohh okay so are we returning that as a valid position in the NMEA postion function? I feel like there should be a check to return the last position if the parser returns NaN

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go (hunk 10 above)
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:311]: it was kind of weird to define nmeamovementsensor with inject. so I kind of opted in for this. I am open to suggestion if there's a better way to go about it.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go (hunk 7 above)
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:328]: instead of testing constructor, I added tests for reconfigure. Didn't seem like there's much to test there.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 12]
```diff
@@ -0,0 +1,328 @@
...
+			NtripURL:             "",
+			NtripConnectAttempts: 10,
+			NtripPass:            "somepass",
+			NtripUser:            "someuser",
+			NtripMountpoint:      "NYC",
+			Board:                testBoardName,
+			I2CBus:               testBusName,
+			I2CAddr:              testi2cAddr,
+		}
+		err := cfg.validateNtrip(path)
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:62]: is it necessary to call the Ntrip and i2c validations each time? I think they will be called when you call the `Validate` function

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 13]
```diff
@@ -0,0 +1,328 @@
...
+	err := g.connect("invalidurl", username, password, 10)
+	test.That(t, err.Error(), test.ShouldContainSubstring, `Can't connect to NTRIP caster`)
+
+	g.ntripClient = makeMockNtripClient()
+
+	err = g.connect(url, username, password, 10)
+	test.That(t, err, test.ShouldBeNil)
+
+	err = g.getStream("", 10)
+	test.That(t, err.Error(), test.ShouldContainSubstring, `lookup fakeurl`)
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:140]: since this function is for testing connect I think there should be a separate `testGetStream` function. also not sure if its worth testing for this random error so if there's not a good way to actually test getting the stream maybe just remove.

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go [hunk 14]
```diff
@@ -0,0 +1,328 @@
...
+	// Position with NaN values.
+	mockSensor.PositionFunc = func(ctx context.Context, extra map[string]interface{}) (*geo.Point, float64, error) {
+		return geo.NewPoint(math.NaN(), math.NaN()), math.NaN(), nil
+	}
+
+	g.lastposition.SetLastPosition(loc1)
+
+	loc3, alt3, err := g.Position(ctx, make(map[string]interface{}))
+	test.That(t, err, test.ShouldBeNil)
+	test.That(t, loc3, test.ShouldResemble, loc1)
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:206]: was a little confused when I first read this line, a comment explaining that it should be returning the last position would be helpful!

File: components/movementsensor/gpsrtkserial/gpsrtkserial_test.go [hunk 15]
```diff
@@ -0,0 +1,238 @@
+package gpsrtkserial
+
+import (
+	"context"
+	"math"
```
Comment (oliviamiller) [inline components/movementsensor/gpsrtkserial/gpsrtkserial_test.go:5]: same comments as the i2c file

**Review by oliviamiller — CHANGES_REQUESTED** [review]

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go (hunk 10 above)
Comment (gvaradarajan) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:312]: fair enough

File: components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go (hunk 13 above)
Comment (susmitaSanyal) [reply components/movementsensor/gpsrtkpmtk/gpsrtkpmtk_test.go:140]: It's really not testing anything. I'll remove it

**Review by oliviamiller — APPROVED** [review]
lgtm!!!!!

**Review by gvaradarajan — APPROVED** [review]
LGTM!
